### PR TITLE
Store rehydration

### PR DESCRIPTION
**Pull request by manzoorwanijk**
Completes #2016 

## Checklist

* [ ] I have read the documentation relating to systems affected by this pull request, see https://github.com/eventespresso/event-espresso-core/tree/master/docs
* [ ] User input is adequately validated and sanitized
* [ ] all publicly displayed strings are internationalized (usually using `esc_html__()`, see https://codex.wordpress.org/I18n_for_WordPress_Developers)
* [ ] My code is tested.
* [ ] My code follows the Event Espresso code style.
* [ ] My code has proper inline documentation.
* [ ] My code accounts for when the site is in Maintenance Mode (MM2 especially disallows usage of models)

### CODE REVIEW

File: assets/prototype/domain/eventEditor/containers/queries/useCacheRehydration.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+				},
+				data: {
+					datetimes,
+				},
+			});
+		}
+		setIsLoaded('datetimes', true);
+
+		const datetimeIn = nodes.map(({ id }) => id);
+		({ nodes = [] } = tickets);
```
Comment (nicolad) [inline assets/prototype/domain/eventEditor/containers/queries/useCacheRehydration.js]: not sure what it does

Comment (manzoorwanijk) [reply assets/prototype/domain/eventEditor/containers/queries/useCacheRehydration.js]: destructuring by using an existing variable i.e. without `let` or `const` when destructuring. You need to wrap it in `()` to make it an expression.

Comment (nicolad) [reply assets/prototype/domain/eventEditor/containers/queries/useCacheRehydration.js]: too much magic involved, can this be accomplished in other ways?

Comment (nicolad) [reply assets/prototype/domain/eventEditor/containers/queries/useCacheRehydration.js]: from what you describe it sounds like the logic might be split into 2 or more small functions

File: assets/prototype/domain/eventEditor/containers/queries/useCacheRehydration.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+import { useApolloClient } from '@apollo/react-hooks';
+import useCacheRehydrationData from './useCacheRehydrationData';
+import useRelations from '../../../../infrastructure/services/relations/useRelations';
+import useStatus from '../../../../infrastructure/services/status/useStatus';
+import useEventId from './useEventId';
+import { queries } from './';
+
+const { GET_TICKETS, GET_DATETIMES, GET_PRICE_TYPES, GET_PRICES } = queries;
+
+const useCacheRehydration = () => {
```
Comment (nicolad) [inline assets/prototype/domain/eventEditor/containers/queries/useCacheRehydration.js]: @manzoorwanijk this isn't exactly the approach I saw here:
https://www.apollographql.com/docs/react/performance/server-side-rendering/#store-rehydration

Comment (nicolad) [reply assets/prototype/domain/eventEditor/containers/queries/useCacheRehydration.js]: my understanding is that this is updating cache on `EventEditor` component mount

Comment (nicolad) [reply assets/prototype/domain/eventEditor/containers/queries/useCacheRehydration.js]: I want to understand why we didn't go through this path: `cache: new InMemoryCache().restore(window.__APOLLO_STATE__)` ?

File: assets/prototype/domain/eventEditor/containers/queries/useCacheRehydration.js (hunk 1 above)
Comment (manzoorwanijk) [reply assets/prototype/domain/eventEditor/containers/queries/useCacheRehydration.js]:
```js
let { name } = person;
```
is same as
```js
let name;
({ name} = person);
```
That way you can reuse an already declared variable

File: assets/prototype/domain/eventEditor/containers/queries/useTicketIds.js [hunk 3]
```diff
@@ -1,35 +1,9 @@
-import * as R from 'ramda';
-import { useApolloClient } from '@apollo/react-hooks';
-
-import useDatetimeIds from './useDatetimeIds';
-import { GET_TICKETS } from './tickets';
-import useStatus from '../../../../infrastructure/services/status/useStatus';
+import useTickets from './useTickets';
 
 const useTicketIds = () => {
```
[the remark this answers is not in the document]
Comment (manzoorwanijk) [reply assets/prototype/domain/eventEditor/containers/queries/useTicketIds.js]: Good idea

File: assets/prototype/domain/eventEditor/containers/queries/useCacheRehydration.js (hunk 2 above)
Comment (manzoorwanijk) [reply assets/prototype/domain/eventEditor/containers/queries/useCacheRehydration.js]: The reason for that is we don't use Apollo Server on the backend. So we need to write to the cache using `client.writeQuery` as mentioned [here](https://www.apollographql.com/docs/react/data/local-state/#initializing-the-cache).

**Pull request merged**